### PR TITLE
Add core files.

### PR DESCRIPTION
**Pull request by mattcollier**
Dependent PR: https://github.com/digitalbazaar/bedrock-authn-did-jwt/pull/2

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+const config = require('bedrock').config;
+
+config['jwt-mongodb'] = {};
+
+// NOTE: config.jsonld.strictSSL from bedrock core should be configured
+// appropriately, the default is `false`
+config['jwt-mongodb'].didio = {
+  // FIXME: some computed config?
+  baseUrl: 'https://authorization.io/dids/'
```
Comment (dlongley) [inline lib/config.js]: Seems like (in the future) we should just create some tiny bedrock package that has some auth.io/DID config stuff in it and then make everyone (who already has their own var) use computed configs based on that.

File: lib/store-hmac.js [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+
+// load config defaults
+require('./config');
+
+const logger = bedrock.loggers.get('app');
+
+// TODO: make responding to this event optional
+bedrock.events.on(
+  'bedrock-authn-did-jwt.config.keyStore', (strategy, callback) =>
+    strategy.registerStore(Store, callback));
```
Comment (dlongley) [inline lib/store-hmac.js]: We should create the `Store` instance here ourselves and pass it to the `strategy`, not pass a class for it to instantiate on our behalf.

File: lib/store-hmac.js [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+  }
+
+  /**
+   * Initialize a namespaced keystore.
+   *
+   * @param options the options to use:
+   *   algorithm the JWT signing algorithm to use.
+   *   clockToleranceInSecs the clockTolerance used during JWT verification.
+   *   namespace the namespace to use.
+   *   ttlInSecs the JWT ttl.
```
Comment (dlongley) [inline lib/store-hmac.js]: Surround any names here that are optional in square brackets, for example:

```js
[ttlInSecs] the JWT ttl.
```

File: lib/store-hmac.js [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+
+    try {
+      decodedToken = jwt.decode(token, {complete: true});
+      header = decodedToken.header;
+    } catch(e) {
+      return callback(e);
+    }
+
+    async.auto({
+      get: callback => this._getState(callback),
```
Comment (dlongley) [inline lib/store-hmac.js]: Would:

```js
get: this._getState.bind(this),
```

Suffice here?

File: lib/store-hmac.js [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+      options = {};
+    }
+    options = options || {};
+    const query = _.assign({}, options.query || {});
+    query.id = database.hash(id);
+    database.collections.jwtKeyStore.update(
+      query, {
+        $set: {
+          'meta.updated': Date.now(),
+          'state': state
```
Comment (dlongley) [inline lib/store-hmac.js]: Unnecessary quotes on LHS `state` ... not sure if linter will complain.

File: lib/store-hmac.js [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+  /**
+   * Gets the cached key state information, retrieving it from the database
+   * and recycling HMAC keys as needed.
+   *
+   * @param options the options to use:
+   *   algorithm the JWT signing algorithm to use.
+   *   clockToleranceInSecs the clockTolerance used during JWT verification.
+   *   namespace the namespace to use.
+   *   provision true for provisioning operation.
+   *   ttlInSecs the JWT ttl.
```
Comment (dlongley) [inline lib/store-hmac.js]: Use square brackets in docs for optional params as needed.

File: lib/store-hmac.js [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+              options.clockToleranceInSecs
+          }
+        };
+
+        this._insert(namespace, state, err => {
+          if(err && err.name === 'DuplicateRecord') {
+            // another process inserted; clear cache, loop, and try again
+            this._keyState[namespace] = null;
+            return process.nextTick(() => {
+              this._getState(namespace, done);
```
Comment (dlongley) [inline lib/store-hmac.js]: `this._getState` doesn't take a `namespace` as its first param. Not sure what should be here, perhaps this:

```js
this._getState({namespace: namespace}, done);
```

?

File: lib/store-hmac.js [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+      collection: 'jwtKeyStore',
+      fields: {id: 1},
+      options: {unique: true, background: false}
+    }], callback)
+  ], callback);
+});
+
+class Store {
+  constructor() {
+    this._keyState = {};
```
Comment (dlongley) [inline lib/store-hmac.js]: Can a `Store` have more than one `namespace`? If not, why is this a map with namespace keys? If so, why is there a single `this.namespace` member?

File: lib/store-hmac.js [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+   * Create a JWT.
+   *
+   * @param payload the payload to sign and include in the JWT.
+   *
+   * @param callback(err, token) called once the operation completes.
+   */
+  sign(payload, callback) {
+    payload = bedrock.util.clone(payload);
+    async.auto({
+      getState: callback => this._getState(callback),
```
Comment (dlongley) [inline lib/store-hmac.js]:
```js
get: this._getState.bind(this);
```
?

File: lib/store-hmac.js [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+      create: ['getState', (callback, results) => {
+        const state = results.getState;
```
Comment (dlongley) [inline lib/store-hmac.js]:
```js
const state = results.get;
```
?

File: lib/store-hmac.js (hunk 2 above)
Comment (dlongley) [reply lib/store-hmac.js]: We should instantiate both an RSA Store and an HMAC Store here -- and create a single object that we pass to the strategy that will handle calling into either one. In other words, that we have two separate stores here is an implementation detail of *this* module, it's not something that the strategy should be getting involved with at all to manage for us. The strategy should expect a simple store object that they can ask to do things without thinking about it.

**Review by dlongley — CHANGES_REQUESTED** [review]
I think there are two primary ways to design this thing and we're sitting somewhere in the middle. Let's chat a bit later about it.

**Comment by mattcollier** [comment]
@dlongley ready for another pass.

File: lib/namespace.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+  verify(options, callback) {
+    callback(new Error('Namespace#verify must be overriden by subclass.'));
+  }
+
+  /**
+   * Create a JWT.
+   *
+   * @param options the options to use:
+   *  payload the payload to sign and include in the JWT.
```
Comment (dlongley) [inline lib/namespace.js]: Should be two-space indent off of the `@`

File: lib/store.js [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+   *     algorithms an array of acceptable HMAC signature algorithms.
+   *     clockToleranceInSecs the acceptable clock skew.
+   * @param callback(err) called once the operation completes.
+   */
+  provision(namespace, options, callback) {
+    let ns;
+    if(namespace === 'RSA') {
+      this.namespaces[namespace] = ns = new NamespaceRsa();
+    } else if(namespace === 'HMAC') {
+      this.namespaces[namespace] = ns = new NamespaceHmac();
```
Comment (dlongley) [inline lib/store.js]: These should be setup just once, not on each `provision` call. Most state information is maintained in the database, not in memory. The only thing in memory in the store should be the interfaces to use the different algorithms -- and we should only need to init those once, when the Store is created. We could init them lazily -- but this code looks like it will create new ones every time a call to `provision` happens with the same algorithm.

File: lib/store.js [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+            updated: now
+          },
+          state: results.createState
+        };
+        database.collections.jwtKeyStore.insert(
+          record, database.writeOptions, (err, result) => {
+            if(err) {
+              if(database.isDuplicateError(err)) {
+                // namespace is already provisioned
+                return callback();
```
Comment (dlongley) [inline lib/store.js]: Need a `TODO: ensure record in database matches given options`. Which may mean that `.insert` should actually be called from the child namespace classes, but we can deal with that later when we deal with the TODO.

File: lib/store.js [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+                err));
+            }
+            callback(null, result.ops[0]);
+          });
+      }]
+    }, callback);
+  }
+
+  sign(options, callback) {
+    async.auto({
```
Comment (dlongley) [inline lib/store.js]: Add `const self = this;`

File: lib/store.js [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+      getState: callback => this.getState(options.namespace, callback),
```
Comment (dlongley) [inline lib/store.js]: Use `self.getState`

File: lib/store.js [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+      sign: ['getState', (callback, results) => {
+        this.namespaces[options.namespace].sign({
```
Comment (dlongley) [inline lib/store.js]: This should be `self.namespaces[options.algorithm]` right? One namespace instance per algorithm? I think we're still having a naming miscommunication because it seems like this design is to create N namespace instances, one per namespace in the database. What we really want is only one of each of these things. Perhaps `Namespace` should be `NamespaceHandler` or `NamespaceType` or `Algorithm`. The algorithm identifiers like `RSA` and `HMAC` each map to a single `NamespaceHandler`/`NamespaceType`/`Algorithm` instance and there are only ever two of these instances right now.

File: lib/store.js [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+      decodedToken = jwt.decode(token, {complete: true});
+      header = decodedToken.header;
+    } catch(err) {
+      return callback(err);
+    }
+    if(header.alg.substring(0, 2) === 'HS') {
+      return async.auto({
+        getState: callback => this.getState('HMAC', callback),
+        verify: ['getState', (callback, results) =>
+          this.namespaces['HMAC'].verify(
```
Comment (dlongley) [inline lib/store.js]: `this.namespaces['HMAC']` => `hmac.verify`

File: lib/store.js [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+    let header;
+    let decodedToken;
+    try {
+      decodedToken = jwt.decode(token, {complete: true});
+      header = decodedToken.header;
+    } catch(err) {
+      return callback(err);
+    }
+    if(header.alg.substring(0, 2) === 'HS') {
+      return async.auto({
```
Comment (dlongley) [inline lib/store.js]: Above return add:

```js
const hmac = this.namespaces.HMAC;
```

File: test/test.config.js [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+const config = require('bedrock').config;
+const path = require('path');
+
+config.mocha.tests.push(path.join(__dirname, 'mocha'));
+
+// MongoDB
+config.mongodb.name = 'bedrock_jwt_mongodb_test';
+config.mongodb.host = 'localhost';
+config.mongodb.port = 27017;
+config.mongodb.local.collection = 'bedrock_jwt_mongodb_test';
```
Comment (dlongley) [inline test/test.config.js]: Do we need to set this anymore? Would be good if we didn't need this extra boilerplate.

File: lib/index.js [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+module.exports = api;
+
+bedrock.events.on(
+  'bedrock-authn-did-jwt.config.keyStore', (strategy, callback) =>
+    strategy.setStore(new Store(), callback));
+
+bedrock.events.on('bedrock-mongodb.ready', callback => {
+  async.waterfall([
+    callback => database.openCollections([
+      'jwtKeyStore', 'jwtKeyStoreHmac'
```
Comment (dlongley) [inline lib/index.js]: Why are there two collections now?

File: lib/namespace-hmac.js [hunk 21]
```diff
@@ -0,0 +1,289 @@
...
+    if(typeof options === 'function') {
+      callback = options;
+      options = {};
+    }
+    const stateId = options.id;
+    const done = callback;
+    async.auto({
+      get: callback => {
+        // use cached key state if available, expiration checked later
+        if(this._keyState) {
```
Comment (dlongley) [inline lib/namespace-hmac.js]: Any time `this` is used inside an `async.auto`, it must be `self`.

File: lib/namespace-hmac.js [hunk 22]
```diff
@@ -0,0 +1,289 @@
...
+          return callback(new BedrockError(
+            'Key not found.',
+            'NotFound',
+            {key: id, httpStatusCode: 404, 'public': true}));
+        }
+        callback(null, record.state, record.meta);
+      });
+  }
+
+  _insert(id, state, callback) {
```
Comment (dlongley) [inline lib/namespace-hmac.js]: It seems like the `insert` was handled in the base class and isn't necessary here. It's also going into a separate collection which I don't understand -- we can chat about it.

**Review by dlongley — CHANGES_REQUESTED** [review]
A few details to discuss.

File: lib/namespace-hmac.js [hunk 23]
```diff
@@ -0,0 +1,287 @@
...
+const logger = bedrock.loggers.get('app');
+
+class NamespaceHmac extends Namespace {
+  constructor() {
+    super();
+    this._keyState = null;
+  }
+
+  createState(options, callback) {
+    const self = this;
```
Comment (davidlehn) [inline lib/namespace-hmac.js]: Is the frequent use of the `self` alias vs plain `this` just for consistency?  The `self` pattern is useful to access an outer `this` from an inner function but many of these uses are on simple functions or ones with inner arrow functions where `self` isn't really needed.

**Comment by mattcollier** [comment]
@dlongley ready for another pass.

File: lib/namespace-handler-hmac.js [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+        }, (err, updated) => {
+          if(err) {
+            logger.error(
+              'Mongo error when trying to get key state.', err);
+            return callback(err);
+          }
+          if(!updated) {
+            // another process updated; clear namespace, loop, and try again
+            logger.verbose(
+              '[jwt-`mongodb`] another process recycled HMAC key.');
```
Comment (dlongley) [inline lib/namespace-handler-hmac.js]: Remove backticks around `mongodb` ... not sure why those are there.

File: lib/namespace-handler-hmac.js [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+            options = _.assign({}, options, {namespace: options.namespace.id});
+            return process.nextTick(() => {
+              self.getKey(options, done);
```
Comment (dlongley) [inline lib/namespace-handler-hmac.js]: Make one-liner:

```js
return process.nextTick(self.getKey.bind(self, options, done));
```

File: lib/namespace-handler-hmac.js [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+            'InternalError',
+            {namespace: id, httpStatusCode: 500, 'public': true},
+            err));
+        }
+        callback(null, result.result.n > 0);
+      });
+  }
+
+  _generateHmacKey() {
+    // TODO: 16 bytes is based on HS256, could be different
```
Comment (dlongley) [inline lib/namespace-handler-hmac.js]: Remove TODO ... I put this here and looking back now it actually makes no sense. 16 bytes is fine, as 128 bits is ridiculously difficult to crack. it's the hash that is 256 bits and this is unrelated. We could make it 32 bits and just never think about it again until someone can harness a few supernovas to crack it. In other words, just remove the TODO. :)

File: lib/namespace-handler-web-key.js [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+      key: options.key
+    }));
+  }
+
+  getKey(options, callback) {
+    const key = options.key;
+    brKey.getPublicKey(key, null, (err, publicKey, meta, privateKey) => {
+      if(err) {
+        return callback(new BedrockError('Invalid signing key specified.',
+          'InvalidKey', {key: key.id}, err));
```
Comment (dlongley) [inline lib/namespace-handler-web-key.js]: `{key: key.id}` => `{key: key}`

File: lib/namespace-handler-web-key.js [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+      }
+      // ensure that the key has not been revoked
+      if(publicKey.sysStatus !== 'active') {
+        return callback(new BedrockError(
+          'The specified signing key has been revoked.',
+          'InvalidKey', {keyId: key.id}));
```
Comment (dlongley) [inline lib/namespace-handler-web-key.js]: `{keyId: key.id}` => `{key: key}`

File: lib/namespace-handler-web-key.js [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+      }
+      callback(
+        null, {id: privateKey.publicKey, material: privateKey.privateKeyPem});
```
Comment (dlongley) [inline lib/namespace-handler-web-key.js]: `{id: key, material: privateKey.privateKeyPem}`

File: lib/namespace-handler.js [hunk 30]
```diff
@@ -0,0 +1,64 @@
...
+   *     state any custom namespace state.
+   *   payload the payload to sign and include in the JWT.
+   *
+   * @param callback(err, token) called once the operation completes.
+   */
+  sign(options, callback) {
+    const self = this;
+    const payload = bedrock.util.clone(options.payload);
+    async.auto({
+      getKey: self.getKey.bind(self, options.namespace.state),
```
Comment (dlongley) [inline lib/namespace-handler.js]:
```js
getKey: self.getKey.bind(self, {namespace: options.namespace})
```

File: lib/namespace-handler.js [hunk 31]
```diff
@@ -0,0 +1,64 @@
...
+      create: ['getKey', (callback, results) => {
+        const key = results.getKey;
+        const nowInSecs = Math.floor(Date.now() / 1000);
+        const notAfter = (nowInSecs + options.namespace.tokenTtlInSecs);
+        _.assign(payload, {exp: notAfter, iat: nowInSecs});
+        const token = jwt.sign(payload, key.material, {
+          algorithm: options.namespace.algorithm,
+          header: {
+            kid: options.namespace.algorithm.startsWith('HS') ?
+              options.namespace.id + ':' + key.id : key.id
```
Comment (dlongley) [inline lib/namespace-handler.js]: Add parens for clarity:

`(options.namespace.id + ':' + key.id) : key.id`

File: lib/store.js [hunk 32]
```diff
@@ -0,0 +1,190 @@
...
+    const handlerWebKey = new NamespaceHandlerWebKey();
+    this.handlers = {
+      HS: new NamespaceHandlerHmac(),
+      RS: handlerWebKey
+      // TODO: Add ECDSA support
+      // ES: handlerWebKey
+    };
+  }
+
+  getNamespace(namespace, callback) {
```
Comment (dlongley) [inline lib/store.js]: Should probably call the param here `id` for better clarity.

File: lib/store.js [hunk 33]
```diff
@@ -0,0 +1,190 @@
...
+    database.collections.jwtKeyStore.findOne(
+      {id: database.hash(namespace)}, {namespace: true}, (err, record) => {
```
Comment (dlongley) [inline lib/store.js]: `database.hash(id)`

File: lib/store.js [hunk 34]
```diff
@@ -0,0 +1,190 @@
...
+        if(err) {
+          logger.error('Mongo error when trying to find key state.', err);
+          return callback(new BedrockError(
+            'Failed to find key state due to internal error.',
+            'InternalError',
+            {namespace: namespace, httpStatusCode: 500, 'public': true}, err));
```
Comment (dlongley) [inline lib/store.js]: `{namespace: id, ...`

File: lib/store.js [hunk 35]
```diff
@@ -0,0 +1,190 @@
...
+        }
+        if(!record) {
+          return callback(new BedrockError(
+            'Key state not found.',
+            'NotFound',
+            {namespace: namespace, httpStatusCode: 404, 'public': true}));
```
Comment (dlongley) [inline lib/store.js]: `{namespace: id, ...`

File: lib/store.js [hunk 36]
```diff
@@ -0,0 +1,190 @@
...
+    try {
+      handler = self._getNamespaceHandler(options.algorithm);
+    } catch(e) {
+      return callback(new BedrockError(
+        'Unsupported algorithm.', 'NotFound',
+        {public: false, namespace: options}));
+    }
+    async.auto({
+      createState: handler.createState.bind(handler, options),
+      storeState: ['createState', (callback, results) => {
```
Comment (dlongley) [inline lib/store.js]: `storeNamespace: ...`

File: lib/store.js [hunk 37]
```diff
@@ -0,0 +1,190 @@
...
+        };
+        database.collections.jwtKeyStore.insert(
+          record, database.writeOptions, (err, result) => {
+            if(err) {
+              if(database.isDuplicateError(err)) {
+                // namespace is already provisioned
+                // TODO: ensure record in database matches given options
+                return callback();
+              }
+              logger.error('Mongo error when trying to insert key state.', err);
```
Comment (dlongley) [inline lib/store.js]: `logger.error('[jwt-mongodb] Mongo error when trying to insert namespace.', err);`

File: lib/store.js [hunk 38]
```diff
@@ -0,0 +1,190 @@
...
+              return callback(new BedrockError(
+                'Failed to insert key state due to internal error.',
```
Comment (dlongley) [inline lib/store.js]: `Failed to insert namespace due ...`

**Review by dlongley — CHANGES_REQUESTED** [review]
Just a few more tweaks and I think we'll finally be there. :)

File: lib/namespace-handler.js (hunk 30 above)
Comment (mattcollier) [reply lib/namespace-handler.js]: this does not work.

**Comment by mattcollier** [comment]
@dlongley ready for another pass.

**Review by dlongley — APPROVED** [review]
